haskell: pkgsStatic fixes

rnhmjoj opened the pull request:
###### Motivation for this change

Fixing common build failures in pkgsStatic.haskellPackages

See the commit messages for details.

###### Things done

- built static GHCs 9.0.2 and 9.2.1
- built `pkgsStatic.haskellPackages.{QuickCheck,pretty-simple}` (previously failing)
- built and tested `pkgsStatic.haskellPackages.pandoc`

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -55,7 +55,8 @@ in
 , doCoverage ? false
-, doHaddock ? !(ghc.isHaLVM or false)
+# See the [HADDOCK_DOCS] note in pkgs/development/compilers/ghc/9.0.2.nix
+, doHaddock ? !(ghc.isHaLVM or false) && !isCross
```
I think a better should would be to add an `hasHaddock` or similar passthru arg to `ghc` and use this here.

sternenseemann commented:
* I'd drop d8dcb588aed526a632c3060c1b091908c7cf31df, since it changes the meaning of `haskell.packages.ghc*` / `haskellPackages` between `pkgs` and `pkgsStatic` which I find problematic. Especially since the compiler build itself is not affected, this doesn't seem necessary?
* Can you add a package to the `staticHaskellPackages` jobset in `pkgs/top-level/release-haskell.nix` which doesn't build without `-fexternal-dynamic-refs`? This would help keeping tabs on it.

rnhmjoj committed:
ghc: enable static RTS

This change allows loading statically compiled libreries into a running
GHC, thus fixing the build of haskell packages that use TemplateHaskell.
See [1] for the details.

Fixes issue #61575, #124284.

[1]: https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/

rnhmjoj committed:
haskellPackages.mkDerivation: check haddock availability

This change makes the haskell builder run the haddockPhase only if the
haddock program is availble (for example, it's not when cross-compiling).

rnhmjoj committed:
release-haskell.nix: test package that needs -fexternal-dynamic-refs

rnhmjoj commented:
> Especially since the compiler build itself is not affected, this doesn't seem necessary?

I see your point, though I'm not sure how useful the default compiler is, given most packages fails to build. Ideally, we should fix the linking errors, but avoiding gmp is the only way I've found so far.

> Can you add a package to the staticHaskellPackages jobset in pkgs/top-level/release-haskell.nix which doesn't build without -fexternal-dynamic-refs? This would help keeping tabs on it.

Done.

cdepillabout commented on pkgs/development/compilers/ghc/9.2.1.nix:
```diff
@@ -107,8 +107,8 @@ let
   '' + lib.optionalString enableRelocatedStaticLibs ''
-    GhcLibHcOpts += -fPIC
-    GhcRtsHcOpts += -fPIC
+    GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
+    GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```
Would you be able to add a comment here why `-fexternal-dynamic-refs` is needed for `enableRelocatedStaticLibs`?

From the GHC manual: https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--fexternal-dynamic-refs

> When generating code, assume that entities imported from a different module might be dynamically linked. This flag is enabled automatically by [-dynamic](https://downloads.haskell.org/ghc/latest/docs/html/users_guide/phases.html#ghc-flag--dynamic).

It is not clear to me why this would be necessary if we are statically linking things.  (Although I don't have a good understanding of static linking, so maybe it should be obvious.)

rnhmjoj replied:
I have to admit I'm baffled by this too and was kind of expecting the haskell maintainers would know better.
Basically, I'm implementing the solution proposed in this blog post: https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/.

cdepillabout replied:
Thanks for the link to the blog post.  It says:

> When evaluating template Haskell splices, GHC will execute compiled expressions in its built-in bytecode interpreter and this code has to be compatible with the RTS of GHC itself. In short, a GHC configured with a dynamic RTS will not be able to load static Haskell libraries to evaluate template Haskell splices.

It then goes on to say:

> We need to build GHC with a static RTS and to make sure that Haskell code is compiled as position independent code so that it can be loaded into a running GHC for template Haskell splices.

Both of these statements make sense of course, but they don't seem to explain why `-fexternal-dynamic-refs` is necessary.

sternenseemann replied:
The GHC man page reads:

> Generate code for linking against dynamic libraries

I assume this means that the RTS will have some kind of dynamic library loading facilities even of static (Haskell?) archives, so the byte code interpreter / whatever used by TH will actually work. `-fPIC` also serves this purpose.

cdepillabout commented on pkgs/development/compilers/ghc/9.0.2.nix:
```diff
@@ -106,8 +106,8 @@ let
     GhcLibWays = "v dyn"
   '' + lib.optionalString enableRelocatedStaticLibs ''
-    GhcLibHcOpts += -fPIC
-    GhcRtsHcOpts += -fPIC
+    GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
+    GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```
Sorry, GitHub isn't letting me create an actual suggestion here, but how about writing this as:

```suggestion
  '' + 
  # These options are recommended by
  # https://www.tweag.io/blog/2020-09-30-bazel-static-haskell/
  # We're not sure why -fexternal-dynamic-refs is necessary.
  lib.optionalString enableRelocatedStaticLibs ''
    GhcLibHcOpts += -fPIC -fexternal-dynamic-refs
    GhcRtsHcOpts += -fPIC -fexternal-dynamic-refs
```

This same change should be made for `9.2.1.nix` as well.

rnhmjoj replied:
Thanks, I added something like this.

cdepillabout reviewed (approved):
Aside from the above comment suggestions, this LGTM!

I tried `nix-build -A pkgsStatic.pretty-simple` and it appears to build and run correctly!

cdepillabout commented:
@sternenseemann Does this look alright to merge to you?

rnhmjoj committed:
pkgsStatic.{haskellPackages,ghc}: default to BigNum

This works around several build failures where programs attempt to link
to the libgmp.so (which is not available), such as:

    Error loading shared library libgmp.so: No such file or directory

rnhmjoj commented:
> but I would be fine with changing the meaning of top level haskellPackages as a compromise,

Looks great to me! Thanks for the suggestion.

sternenseemann commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12445,7 +12449,11 @@ with pkgs;
   # however, targetPackages won't be populated, so we need to fall back to the
   # plain, cross-compiled compiler (which is only theoretical at the moment).
-  ghc = targetPackages.haskellPackages.ghc or haskell.compiler.ghc902;
+  ghc = dontRecurseIntoAttrs
```
The old `targetPackages.haskellPackages.ghc or …` needs to be preserved, since it's not the same as `haskell.compiler.ghc902` (e.g. it has a `withPackages` method).

rnhmjoj replied:
I messed up by copying one too many line from the other change. Thanks.

cdepillabout commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -55,7 +55,7 @@ in
 , mainProgram ? null
 , doCoverage ? false
-, doHaddock ? !(ghc.isHaLVM or false)
+, doHaddock ? !(ghc.isHaLVM or false) && ghc.hasHaddock
```
I think this suggestion makes sense.  But on the other hand, as an end user, if I pass `doHaddock = true`, I'd expect either I get haddocks when the build finishes, or the build fails.  It could be somewhat confusing to specify `doHaddock = true`, but have the build successfully finish without building haddocks.

Either way, I'd say this potential fix probably shouldn't hold up merging in this PR.

cdepillabout commented:
> I would be fine with changing the meaning of top level haskellPackages as a compromise

I agree this was a good suggestion.  I think the current state of this PR makes a lot of sense, having `pkgsStatic.haskellPackages` explicitly pointing at `pkgsStatic.haskell.packages.native-bignum.ghc-902`.

> Have you tried actually building anything that uses FFI in TH? I guess that would need a dynamically linked library as well?

That's a good question.  Do you happen to know a package that uses FFI in TH off the top of your head?  I gave it some thought, but I couldn't come up with anything.  (I did try compiling `pkgsStatic.haskellPackages.zlib`, which does successfully compile, but I imagine it is unlikely to use TH.)

Although even if this doesn't work, I don't feel like it should hold up merging this PR, since this PR currently improves the status quo.

rnhmjoj commented:
So, can we merge this?

sternenseemann commented:
Probably, I'll try to have a final look soon. A small request: Could you use `haskellPackages.mkDerivation` over `haskell-generic-builder` in the commit message? I'd like to have correspondence to an attribute where possible. Rebasing on haskell-updates would be useful anyways, so CI wouldn't be broken by an unrelated problem.

rnhmjoj commented:
Commit message updated and rebased.

rnhmjoj commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12490,7 +12494,11 @@ with pkgs;
   # however, targetPackages won't be populated, so we need to fall back to the
   # plain, cross-compiled compiler (which is only theoretical at the moment).
-  ghc = targetPackages.haskellPackages.ghc or haskell.compiler.ghc902;
+  ghc = dontRecurseIntoAttrs
```
It seems so. Fixed, again.

sternenseemann commented on pkgs/development/haskell-modules/generic-builder.nix:
```diff
@@ -55,7 +55,7 @@ in
 , mainProgram ? null
 , doCoverage ? false
-, doHaddock ? !(ghc.isHaLVM or false)
+, doHaddock ? !(ghc.isHaLVM or false) && ghc.hasHaddock
```
The remaining GHCs (8.8.4, 8.10.7, HEAD) need this annotation as well, e. g. `ghc*Binary`, but also `ghcjs`, see the current eval error as per ofborg.

rnhmjoj replied:
The binary compilers don't have the `enableHaddockProgram` arguments. What should I set `hasHaddock` to?

rnhmjoj replied:
What about just doing `&& (ghc.hasHaddock or true)`?

sternenseemann replied:
> What about just doing `&& (ghc.hasHaddock or true)`?

Also works and probably a smart move since it means that we don't have to commit to having this attribute forever.

rnhmjoj replied:
Ok, did it this way. I also added `hasHaddock` to `ghcHEAD`, so the main versions are covered.

rnhmjoj replied:
The evaluation is fixed, now.

sternenseemann replied:
GHC 8.8.4 and 8.10.7 also have an `enableHaddockProgram` argument, don't they?

rnhmjoj replied:
Yes, but I though it would be enough to target the next major version 9.0 and head.
Do you want me to add those too?

sternenseemann replied:
Yeah, definitely, the package sets are still around and work for the most part and they should behave like the default ones as much as possible!

rnhmjoj replied:
Done.

rnhmjoj commented:
@sternenseemann Thank you for reviewing!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/statically-linking-haskell-fails-to-link-libffi/18317/2

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/statically-linking-haskell-fails-to-link-libffi/18317/3
